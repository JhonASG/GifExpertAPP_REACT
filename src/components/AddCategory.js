import React, {useState} from "react";
import PropsTypes from 'prop-types';

const AddCategory = ({setCategories}) =>{
    // Obtener el valor del input y voy a modificarle el mismo
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(cat => [inputValue, ...cat]);
            setInputValue('');
        }
    }

    return (
        /*En este caso no se necesita un fragment <></> porque con el <form></form> 
        agrupamos todos los elementos.*/
        <form onSubmit={handleSubmit}>
            <input type="text" value={inputValue} onChange={handleInputChange}/>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropsTypes.func.isRequired
}

export default AddCategory;