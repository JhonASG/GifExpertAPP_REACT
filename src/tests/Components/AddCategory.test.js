import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import AddCategory from "../../components/AddCategory";

describe('Pruebas en el componente Add Category', () => {
    // jest.fn(), simula una función
	const setCategories = jest.fn();
    let view;

    let renderComponent = () => {
        view = render(<AddCategory setCategories={setCategories}/>);
    };

    beforeEach(() => {
		jest.clearAllMocks();
        renderComponent();
    })
    
    test('Debe renderizarse correctamente el componente', () => {
        expect(view).toMatchSnapshot();
    });

    test('Debe de cambiar la caja de texto', () => {
		const input = screen.getByRole('textbox');
		const value = "Hola Mundo";

		fireEvent.change(input, {target: {value: value}});

		const valueInput = input.getAttribute('value'); 
		expect(valueInput).toBe(value);
    });

	test('No debe de postear la información con submit', () => {
		const form = screen.getByTestId('form');

		fireEvent.submit(form, { preventDefault(){} });

		// Es verificar cuando al no hemos pasado ningún valor al evento handleInputChange
		expect( setCategories ).not.toHaveBeenCalled();
	});

	test('debe de llama el setCategories y limpiar la caja de texto', () => {
		// Simular inputChange
		const input = screen.getByRole('textbox');
		const value = "Goku";

		fireEvent.change(input, {target: {value: value}});

		// Simular Submit
		const form = screen.getByTestId('form');

		fireEvent.submit(form, { preventDefault(){} });

		// setCategories se debe de haber llamado
		expect(setCategories).toHaveBeenCalled();
		// Verificar que setCategories se halla llamado como una función
		expect(setCategories).toHaveBeenCalledWith( expect.any(Function) );

		// inputValue debe estar '' (vacio)
		const valueInput = input.getAttribute('value');
		expect(valueInput).toBe('');
	});
});