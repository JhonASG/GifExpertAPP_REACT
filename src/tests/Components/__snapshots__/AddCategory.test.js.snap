// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente Add Category Debe renderizarse correctamente el componente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        data-testid="form"
      >
        <input
          type="text"
          value=""
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      data-testid="form"
    >
      <input
        type="text"
        value=""
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
